uxsim/uxsim.py

```diff
@@ -952,51 +953,56 @@ def __init__(s, W):
+        s.pred = np.zeros([len(s.W.NODES), len(s.W.NODES)], dtype=int)
+
+        # For homogeneous DUO, 1 if on the shortest path to k
+        s.route_pref = {k.id: {l: 0 for l in s.W.LINKS} for k in s.W.NODES}
 
     def route_search_all(s, infty=np.inf, noise=0):
         """
-        Compute the current shortest path based on instantanious travel time.
+        Compute the current shortest path based on instantaneous travel time.
 
         Parameters
         ----------
         infty : float
             value representing infinity.
         noise : float
-            very small noise to slightly randomize route choice. useful to eliminate strange results at an initial stage of simulation where many routes has identical travel time.
+            very small noise to slightly randomize route choice. Useful to eliminate strange results at an initial stage of simulation where many routes have identical travel time.
         """
         for link in s.W.LINKS:
             i = link.start_node.id
             j = link.end_node.id
-            if s.W.ADJ_MAT[i,j]:
-                s.adj_mat_time[i,j] = link.traveltime_instant[-1]*random.uniform(1, 1+noise) + link.route_choice_penalty
-                if link.capacity_in == 0: #流入禁止の場合は通行不可
-                    s.adj_mat_time[i,j] = np.inf
+            if s.W.ADJ_MAT[i, j]:
+                s.adj_mat_time[i, j] = link.traveltime_instant[-1] * np.random.uniform(1, 1 + noise) + link.route_choice_penalty
+                if link.capacity_in == 0:  # If entry is prohibited, then passage is not possible
+                    s.adj_mat_time[i, j] = np.inf
             else:
-                s.adj_mat_time[i,j] = np.inf
+                s.adj_mat_time[i, j] = np.inf
 
-        s.dist, s.pred = floyd_warshall(s.adj_mat_time, return_predecessors=True)
+        # Convert adjacency matrix time to CSR format for efficient graph analysis
+        csr_mat = csr_matrix(s.adj_mat_time)
+        # Using Dijkstra's algorithm to find the shortest paths
+        s.dist, s.pred = dijkstra(csgraph=csr_mat, directed=True, indices=None, return_predecessors=True)
 
+        # Directly use the predecessors to update the next matrix for route choice
+        # Vectorized update of s.next
         n_vertices = s.pred.shape[0]
-        s.next = -np.ones((n_vertices, n_vertices), dtype=int)
-        for i in range(n_vertices):
-            for j in range(n_vertices):
-                # iからjへの最短経路を逆にたどる．．． -> todo: 起終点を逆にした最短経路探索にすればよい
-                if i != j:
-                    prev = j
-                    while s.pred[i, prev] != i and s.pred[i, prev] != -9999:
-                        prev = s.pred[i, prev]
-                    s.next[i, j] = prev
+        s.next = np.full((n_vertices, n_vertices), -1, dtype=int)  # Initialize with -1
+
+        # Vectorized operation to copy `s.pred` into `s.next` where pred != -9999
+        valid_pred = s.pred != -9999
+        s.next[valid_pred] = s.pred[valid_pred]
```

Comment:
Is this correct? I think `pred` and `next` are different concept and cannot be mapped like this.